Readme.md
# Buffer

Generic buffer for batching entries, such as log entries.

---

[![GoDoc](https://godoc.org/github.com/tj/go-buffer?status.svg)](https://godoc.org/github.com/tj/go-buffer)
![](https://img.shields.io/badge/license-MIT-blue.svg)
![](https://img.shields.io/badge/status-stable-green.svg)
![](https://github.com/tj/go-buffer/workflows/Tests/badge.svg)

## Sponsors

This project is sponsored by [CTO.ai](https://cto.ai/), making it easy for development teams to create and share workflow automations without leaving the command line.

[![](https://apex-software.imgix.net/github/sponsors/cto.png)](https://cto.ai/)

And my [GitHub sponsors](https://github.com/sponsors/tj):

[<img src="https://sponsors-api-u2fftug6kq-uc.a.run.app/sponsor/avatar/0" width="35">](https://sponsors-api-u2fftug6kq-uc.a.run.app/sponsor/profile/0)
[<img src="https://sponsors-api-u2fftug6kq-uc.a.run.app/sponsor/avatar/1" width="35">](https://sponsors-api-u2fftug6kq-uc.a.run.app/sponsor/profile/1)
[<img src="https://sponsors-api-u2fftug6kq-uc.a.run.app/sponsor/avatar/2" width="35">](https://sponsors-api-u2fftug6kq-uc.a.run.app/sponsor/profile/2)
[<img src="https://sponsors-api-u2fftug6kq-uc.a.run.app/sponsor/avatar/3" width="35">](https://sponsors-api-u2fftug6kq-uc.a.run.app/sponsor/profile/3)
[<img src="https://sponsors-api-u2fftug6kq-uc.a.run.app/sponsor/avatar/4" width="35">](https://sponsors-api-u2fftug6kq-uc.a.run.app/sponsor/profile/4)
[<img src="https://sponsors-api-u2fftug6kq-uc.a.run.app/sponsor/avatar/5" width="35">](https://sponsors-api-u2fftug6kq-uc.a.run.app/sponsor/profile/5)
[<img src="https://sponsors-api-u2fftug6kq-uc.a.run.app/sponsor/avatar/6" width="35">](https://sponsors-api-u2fftug6kq-uc.a.run.app/sponsor/profile/6)
[<img src="https://sponsors-api-u2fftug6kq-uc.a.run.app/sponsor/avatar/7" width="35">](https://sponsors-api-u2fftug6kq-uc.a.run.app/sponsor/profile/7)
[<img src="https://sponsors-api-u2fftug6kq-uc.a.run.app/sponsor/avatar/8" width="35">](https://sponsors-api-u2fftug6kq-uc.a.run.app/sponsor/profile/8)
[<img src="https://sponsors-api-u2fftug6kq-uc.a.run.app/sponsor/avatar/9" width="35">](https://sponsors-api-u2fftug6kq-uc.a.run.app/sponsor/profile/9)
[<img src="https://sponsors-api-u2fftug6kq-uc.a.run.app/sponsor/avatar/10" width="35">](https://sponsors-api-u2fftug6kq-uc.a.run.app/sponsor/profile/10)
[<img src="https://sponsors-api-u2fftug6kq-uc.a.run.app/sponsor/avatar/11" width="35">](https://sponsors-api-u2fftug6kq-uc.a.run.app/sponsor/profile/11)
[<img src="https://sponsors-api-u2fftug6kq-uc.a.run.app/sponsor/avatar/12" width="35">](https://sponsors-api-u2fftug6kq-uc.a.run.app/sponsor/profile/12)
[<img src="https://sponsors-api-u2fftug6kq-uc.a.run.app/sponsor/avatar/13" width="35">](https://sponsors-api-u2fftug6kq-uc.a.run.app/sponsor/profile/13)
[<img src="https://sponsors-api-u2fftug6kq-uc.a.run.app/sponsor/avatar/14" width="35">](https://sponsors-api-u2fftug6kq-uc.a.run.app/sponsor/profile/14)
[<img src="https://sponsors-api-u2fftug6kq-uc.a.run.app/sponsor/avatar/15" width="35">](https://sponsors-api-u2fftug6kq-uc.a.run.app/sponsor/profile/15)
[<img src="https://sponsors-api-u2fftug6kq-uc.a.run.app/sponsor/avatar/16" width="35">](https://sponsors-api-u2fftug6kq-uc.a.run.app/sponsor/profile/16)
[<img src="https://sponsors-api-u2fftug6kq-uc.a.run.app/sponsor/avatar/17" width="35">](https://sponsors-api-u2fftug6kq-uc.a.run.app/sponsor/profile/17)
[<img src="https://sponsors-api-u2fftug6kq-uc.a.run.app/sponsor/avatar/18" width="35">](https://sponsors-api-u2fftug6kq-uc.a.run.app/sponsor/profile/18)
[<img src="https://sponsors-api-u2fftug6kq-uc.a.run.app/sponsor/avatar/19" width="35">](https://sponsors-api-u2fftug6kq-uc.a.run.app/sponsor/profile/19)
[<img src="https://sponsors-api-u2fftug6kq-uc.a.run.app/sponsor/avatar/20" width="35">](https://sponsors-api-u2fftug6kq-uc.a.run.app/sponsor/profile/20)
[<img src="https://sponsors-api-u2fftug6kq-uc.a.run.app/sponsor/avatar/21" width="35">](https://sponsors-api-u2fftug6kq-uc.a.run.app/sponsor/profile/21)
[<img src="https://sponsors-api-u2fftug6kq-uc.a.run.app/sponsor/avatar/22" width="35">](https://sponsors-api-u2fftug6kq-uc.a.run.app/sponsor/profile/22)
[<img src="https://sponsors-api-u2fftug6kq-uc.a.run.app/sponsor/avatar/23" width="35">](https://sponsors-api-u2fftug6kq-uc.a.run.app/sponsor/profile/23)
[<img src="https://sponsors-api-u2fftug6kq-uc.a.run.app/sponsor/avatar/24" width="35">](https://sponsors-api-u2fftug6kq-uc.a.run.app/sponsor/profile/24)
[<img src="https://sponsors-api-u2fftug6kq-uc.a.run.app/sponsor/avatar/25" width="35">](https://sponsors-api-u2fftug6kq-uc.a.run.app/sponsor/profile/25)
[<img src="https://sponsors-api-u2fftug6kq-uc.a.run.app/sponsor/avatar/26" width="35">](https://sponsors-api-u2fftug6kq-uc.a.run.app/sponsor/profile/26)
[<img src="https://sponsors-api-u2fftug6kq-uc.a.run.app/sponsor/avatar/27" width="35">](https://sponsors-api-u2fftug6kq-uc.a.run.app/sponsor/profile/27)
[<img src="https://sponsors-api-u2fftug6kq-uc.a.run.app/sponsor/avatar/28" width="35">](https://sponsors-api-u2fftug6kq-uc.a.run.app/sponsor/profile/28)
[<img src="https://sponsors-api-u2fftug6kq-uc.a.run.app/sponsor/avatar/29" width="35">](https://sponsors-api-u2fftug6kq-uc.a.run.app/sponsor/profile/29)
[<img src="https://sponsors-api-u2fftug6kq-uc.a.run.app/sponsor/avatar/30" width="35">](https://sponsors-api-u2fftug6kq-uc.a.run.app/sponsor/profile/30)
[<img src="https://sponsors-api-u2fftug6kq-uc.a.run.app/sponsor/avatar/31" width="35">](https://sponsors-api-u2fftug6kq-uc.a.run.app/sponsor/profile/31)
[<img src="https://sponsors-api-u2fftug6kq-uc.a.run.app/sponsor/avatar/32" width="35">](https://sponsors-api-u2fftug6kq-uc.a.run.app/sponsor/profile/32)
[<img src="https://sponsors-api-u2fftug6kq-uc.a.run.app/sponsor/avatar/33" width="35">](https://sponsors-api-u2fftug6kq-uc.a.run.app/sponsor/profile/33)
[<img src="https://sponsors-api-u2fftug6kq-uc.a.run.app/sponsor/avatar/34" width="35">](https://sponsors-api-u2fftug6kq-uc.a.run.app/sponsor/profile/34)
[<img src="https://sponsors-api-u2fftug6kq-uc.a.run.app/sponsor/avatar/35" width="35">](https://sponsors-api-u2fftug6kq-uc.a.run.app/sponsor/profile/35)
[<img src="https://sponsors-api-u2fftug6kq-uc.a.run.app/sponsor/avatar/36" width="35">](https://sponsors-api-u2fftug6kq-uc.a.run.app/sponsor/profile/36)
[<img src="https://sponsors-api-u2fftug6kq-uc.a.run.app/sponsor/avatar/37" width="35">](https://sponsors-api-u2fftug6kq-uc.a.run.app/sponsor/profile/37)
[<img src="https://sponsors-api-u2fftug6kq-uc.a.run.app/sponsor/avatar/38" width="35">](https://sponsors-api-u2fftug6kq-uc.a.run.app/sponsor/profile/38)
[<img src="https://sponsors-api-u2fftug6kq-uc.a.run.app/sponsor/avatar/39" width="35">](https://sponsors-api-u2fftug6kq-uc.a.run.app/sponsor/profile/39)
[<img src="https://sponsors-api-u2fftug6kq-uc.a.run.app/sponsor/avatar/40" width="35">](https://sponsors-api-u2fftug6kq-uc.a.run.app/sponsor/profile/40)
[<img src="https://sponsors-api-u2fftug6kq-uc.a.run.app/sponsor/avatar/41" width="35">](https://sponsors-api-u2fftug6kq-uc.a.run.app/sponsor/profile/41)
[<img src="https://sponsors-api-u2fftug6kq-uc.a.run.app/sponsor/avatar/42" width="35">](https://sponsors-api-u2fftug6kq-uc.a.run.app/sponsor/profile/42)
[<img src="https://sponsors-api-u2fftug6kq-uc.a.run.app/sponsor/avatar/43" width="35">](https://sponsors-api-u2fftug6kq-uc.a.run.app/sponsor/profile/43)
[<img src="https://sponsors-api-u2fftug6kq-uc.a.run.app/sponsor/avatar/44" width="35">](https://sponsors-api-u2fftug6kq-uc.a.run.app/sponsor/profile/44)
[<img src="https://sponsors-api-u2fftug6kq-uc.a.run.app/sponsor/avatar/45" width="35">](https://sponsors-api-u2fftug6kq-uc.a.run.app/sponsor/profile/45)
[<img src="https://sponsors-api-u2fftug6kq-uc.a.run.app/sponsor/avatar/46" width="35">](https://sponsors-api-u2fftug6kq-uc.a.run.app/sponsor/profile/46)
[<img src="https://sponsors-api-u2fftug6kq-uc.a.run.app/sponsor/avatar/47" width="35">](https://sponsors-api-u2fftug6kq-uc.a.run.app/sponsor/profile/47)
[<img src="https://sponsors-api-u2fftug6kq-uc.a.run.app/sponsor/avatar/48" width="35">](https://sponsors-api-u2fftug6kq-uc.a.run.app/sponsor/profile/48)
[<img src="https://sponsors-api-u2fftug6kq-uc.a.run.app/sponsor/avatar/49" width="35">](https://sponsors-api-u2fftug6kq-uc.a.run.app/sponsor/profile/49)
[<img src="https://sponsors-api-u2fftug6kq-uc.a.run.app/sponsor/avatar/50" width="35">](https://sponsors-api-u2fftug6kq-uc.a.run.app/sponsor/profile/50)
[<img src="https://sponsors-api-u2fftug6kq-uc.a.run.app/sponsor/avatar/51" width="35">](https://sponsors-api-u2fftug6kq-uc.a.run.app/sponsor/profile/51)
[<img src="https://sponsors-api-u2fftug6kq-uc.a.run.app/sponsor/avatar/52" width="35">](https://sponsors-api-u2fftug6kq-uc.a.run.app/sponsor/profile/52)
[<img src="https://sponsors-api-u2fftug6kq-uc.a.run.app/sponsor/avatar/53" width="35">](https://sponsors-api-u2fftug6kq-uc.a.run.app/sponsor/profile/53)
[<img src="https://sponsors-api-u2fftug6kq-uc.a.run.app/sponsor/avatar/54" width="35">](https://sponsors-api-u2fftug6kq-uc.a.run.app/sponsor/profile/54)
[<img src="https://sponsors-api-u2fftug6kq-uc.a.run.app/sponsor/avatar/55" width="35">](https://sponsors-api-u2fftug6kq-uc.a.run.app/sponsor/profile/55)
[<img src="https://sponsors-api-u2fftug6kq-uc.a.run.app/sponsor/avatar/56" width="35">](https://sponsors-api-u2fftug6kq-uc.a.run.app/sponsor/profile/56)
[<img src="https://sponsors-api-u2fftug6kq-uc.a.run.app/sponsor/avatar/57" width="35">](https://sponsors-api-u2fftug6kq-uc.a.run.app/sponsor/profile/57)
[<img src="https://sponsors-api-u2fftug6kq-uc.a.run.app/sponsor/avatar/58" width="35">](https://sponsors-api-u2fftug6kq-uc.a.run.app/sponsor/profile/58)
[<img src="https://sponsors-api-u2fftug6kq-uc.a.run.app/sponsor/avatar/59" width="35">](https://sponsors-api-u2fftug6kq-uc.a.run.app/sponsor/profile/59)
[<img src="https://sponsors-api-u2fftug6kq-uc.a.run.app/sponsor/avatar/60" width="35">](https://sponsors-api-u2fftug6kq-uc.a.run.app/sponsor/profile/60)
[<img src="https://sponsors-api-u2fftug6kq-uc.a.run.app/sponsor/avatar/61" width="35">](https://sponsors-api-u2fftug6kq-uc.a.run.app/sponsor/profile/61)
[<img src="https://sponsors-api-u2fftug6kq-uc.a.run.app/sponsor/avatar/62" width="35">](https://sponsors-api-u2fftug6kq-uc.a.run.app/sponsor/profile/62)
[<img src="https://sponsors-api-u2fftug6kq-uc.a.run.app/sponsor/avatar/63" width="35">](https://sponsors-api-u2fftug6kq-uc.a.run.app/sponsor/profile/63)
[<img src="https://sponsors-api-u2fftug6kq-uc.a.run.app/sponsor/avatar/64" width="35">](https://sponsors-api-u2fftug6kq-uc.a.run.app/sponsor/profile/64)
[<img src="https://sponsors-api-u2fftug6kq-uc.a.run.app/sponsor/avatar/65" width="35">](https://sponsors-api-u2fftug6kq-uc.a.run.app/sponsor/profile/65)
[<img src="https://sponsors-api-u2fftug6kq-uc.a.run.app/sponsor/avatar/66" width="35">](https://sponsors-api-u2fftug6kq-uc.a.run.app/sponsor/profile/66)
[<img src="https://sponsors-api-u2fftug6kq-uc.a.run.app/sponsor/avatar/67" width="35">](https://sponsors-api-u2fftug6kq-uc.a.run.app/sponsor/profile/67)
[<img src="https://sponsors-api-u2fftug6kq-uc.a.run.app/sponsor/avatar/68" width="35">](https://sponsors-api-u2fftug6kq-uc.a.run.app/sponsor/profile/68)


History.md

v1.2.0 / 2020-08-06
===================

  * change DefaultFlushInterval to 30s

v1.1.0 / 2020-08-03
===================

  * add Buffer.FlushSync() for synchronous flushing


buffer_test.go
package buffer_test

import (
	"context"
	"fmt"
	"sync"
	"testing"
	"time"

	"github.com/tj/assert"

	"github.com/tj/go-buffer"
)

// Test forced flush.
func TestBuffer_Flush(t *testing.T) {
	var mu sync.Mutex
	var flushes int
	var flushed []interface{}

	flush := func(ctx context.Context, values []interface{}) error {
		mu.Lock()
		defer mu.Unlock()
		flushes++
		flushed = append(flushed, values...)
		return nil
	}

	errors := func(err error) {
		assert.NoError(t, err)
	}

	b := buffer.New(
		buffer.WithFlushHandler(flush),
		buffer.WithErrorHandler(errors),
	)

	b.Push("hello")
	b.Push(" ")

	b.Flush()

	b.Push("world")
	b.Push("!")

	b.Close()

	assert.Equal(t, 2, flushes, "flush count")
	assert.Len(t, flushed, 4)
}

// Test sync forced flush.
func TestBuffer_FlushSync(t *testing.T) {
	var mu sync.Mutex
	var flushes int
	var flushed []interface{}

	flush := func(ctx context.Context, values []interface{}) error {
		mu.Lock()
		defer mu.Unlock()
		flushes++
		flushed = append(flushed, values...)
		return nil
	}

	errors := func(err error) {
		assert.NoError(t, err)
	}

	b := buffer.New(
		buffer.WithFlushHandler(flush),
		buffer.WithErrorHandler(errors),
	)

	b.Push("hello")
	b.Push(" ")

	b.FlushSync()

	b.Push("world")
	b.Push("!")

	b.Close()

	assert.Equal(t, 2, flushes, "flush count")
	assert.Len(t, flushed, 4)
}

// Test max entries.
func TestBuffer_maxEntries(t *testing.T) {
	var mu sync.Mutex
	var flushes int
	var flushed []interface{}

	flush := func(ctx context.Context, values []interface{}) error {
		mu.Lock()
		defer mu.Unlock()
		flushes++
		flushed = append(flushed, values...)
		return nil
	}

	errors := func(err error) {
		assert.NoError(t, err)
	}

	b := buffer.New(
		buffer.WithFlushHandler(flush),
		buffer.WithErrorHandler(errors),
		buffer.WithMaxEntries(3),
	)

	b.Push("hello")
	b.Push(" ")
	b.Push("world")
	b.Push("!")

	b.Close()

	assert.Len(t, flushed, 4)
	assert.Equal(t, 2, flushes, "flush count")
}

// Test flush interval.
func TestBuffer_flushInterval(t *testing.T) {
	var mu sync.Mutex
	var flushes int
	var flushed []interface{}

	flush := func(ctx context.Context, values []interface{}) error {
		mu.Lock()
		defer mu.Unlock()
		flushes++
		flushed = append(flushed, values...)
		return nil
	}

	errors := func(err error) {
		assert.NoError(t, err)
	}

	b := buffer.New(
		buffer.WithFlushHandler(flush),
		buffer.WithErrorHandler(errors),
		buffer.WithMaxEntries(100),
		buffer.WithFlushInterval(time.Millisecond*50),
	)

	b.Push("hello")
	time.Sleep(time.Millisecond * 150)

	b.Push(" ")
	time.Sleep(time.Millisecond * 150)

	b.Push("world")
	b.Push("!")

	b.Close()

	assert.Len(t, flushed, 4)
	assert.Equal(t, 3, flushes, "flush count")
}

// temporaryError .
type temporaryError struct{}

// Error implementation.
func (e temporaryError) Error() string {
	return "rate limited"
}

// Temporary implementation.
func (e temporaryError) Temporary() bool {
	return true
}

// Test flush retries on temporary errors.
func TestBuffer_flushRetriesOk(t *testing.T) {
	var mu sync.Mutex
	var flushes int

	flush := func(ctx context.Context, values []interface{}) error {
		mu.Lock()
		defer mu.Unlock()
		flushes++
		if flushes < 3 {
			return temporaryError{}
		}
		return nil
	}

	errors := func(err error) {
		assert.NoError(t, err)
	}

	b := buffer.New(
		buffer.WithFlushHandler(flush),
		buffer.WithErrorHandler(errors),
	)

	b.Push("hello")
	b.Push("world")
	b.Push("!")

	b.Close()

	assert.Equal(t, 3, flushes, "flush count")
}

// Test exceeding retries.
func TestBuffer_flushRetriesExceeded(t *testing.T) {
	var mu sync.Mutex
	var flushes int
	var err error

	flush := func(ctx context.Context, values []interface{}) error {
		mu.Lock()
		defer mu.Unlock()
		flushes++
		return temporaryError{}
	}

	errors := func(e error) {
		err = e
	}

	b := buffer.New(
		buffer.WithFlushHandler(flush),
		buffer.WithErrorHandler(errors),
	)

	b.Push("hello")
	b.Push("world")
	b.Push("!")

	b.Close()

	assert.Equal(t, 3, flushes, "flush count")
	assert.EqualError(t, err, `rate limited`)
}

// Test regular flush errors.
func TestBuffer_flushErrors(t *testing.T) {
	var mu sync.Mutex
	var flushes int
	var err error

	flush := func(ctx context.Context, values []interface{}) error {
		mu.Lock()
		defer mu.Unlock()
		flushes++
		return fmt.Errorf("boom")
	}

	errors := func(e error) {
		err = e
	}

	b := buffer.New(
		buffer.WithFlushHandler(flush),
		buffer.WithErrorHandler(errors),
	)

	b.Push("hello")
	b.Push("world")
	b.Push("!")

	b.Close()

	assert.Equal(t, 1, flushes, "flush count")
	assert.EqualError(t, err, `boom`)
}

// Benchmark pushing.
func BenchmarkPush(b *testing.B) {
	buf := buffer.New(
		buffer.WithMaxEntries(250),
	)

	b.ResetTimer()
	b.ReportAllocs()

	for i := 0; i < b.N; i++ {
		buf.Push("hello")
	}

	buf.Close()
}


buffer.go
// Package buffer provides a generic buffer or batching mechanism for flushing
// entries at a given size or interval, useful for cases such as batching log
// events.
package buffer

import (
	"context"
	"log"
	"os"
	"sync"
	"time"
)

// TODO: make logging optional, accept a Printfer

// logs instance.
var logs = log.New(os.Stderr, "buffer ", log.LstdFlags)

// temporary is the interfaced used for temporary errors.
type temporary interface {
	Temporary() bool
}

// DefaultMaxEntries is the default max entries limit.
var DefaultMaxEntries = 250

// DefaultMaxRetries is the default max retries limit.
var DefaultMaxRetries = 3

// DefaultFlushInterval is the default flush interval.
var DefaultFlushInterval = time.Second * 30

// DefaultFlushTimeout is the default flush timeout.
var DefaultFlushTimeout = time.Second * 15

// FlushFunc is the flush callback function used to flush entries.
type FlushFunc func(context.Context, []interface{}) error

// ErrorFunc is the error callback function used to report flushing errors,
// called when flushing fails and MaxRetries has exceeded on temporary errors.
type ErrorFunc func(error)

// Option function.
type Option func(*Buffer)

// New buffer with the given options.
func New(options ...Option) *Buffer {
	var v Buffer
	v.handleFlush = noopFlush
	v.handleError = noopError
	v.maxEntries = DefaultMaxEntries
	v.maxRetries = DefaultMaxRetries
	v.flushInterval = DefaultFlushInterval
	v.flushTimeout = DefaultFlushTimeout

	for _, o := range options {
		o(&v)
	}

	v.done = make(chan struct{})
	if v.flushInterval > 0 {
		go v.intervalFlush()
	}

	return &v
}

// WithFlushHandler sets the function handling flushes.
func WithFlushHandler(fn FlushFunc) Option {
	return func(v *Buffer) {
		v.handleFlush = fn
	}
}

// WithErrorHandler sets the function handling errors.
func WithErrorHandler(fn ErrorFunc) Option {
	return func(v *Buffer) {
		v.handleError = fn
	}
}

// WithMaxEntries sets the maximum number of entries before flushing.
func WithMaxEntries(n int) Option {
	return func(v *Buffer) {
		v.maxEntries = n
	}
}

// WithMaxRetries sets the maximum number of retries for temporary flush errors.
func WithMaxRetries(n int) Option {
	return func(v *Buffer) {
		v.maxRetries = n
	}
}

// WithFlushInterval sets the interval at which events are periodically flushed.
func WithFlushInterval(d time.Duration) Option {
	return func(v *Buffer) {
		v.flushInterval = d
	}
}

// WithFlushTimeout sets the flush timeout.
func WithFlushTimeout(d time.Duration) Option {
	return func(v *Buffer) {
		v.flushTimeout = d
	}
}

// Buffer is used to batch entries.
type Buffer struct {
	pendingFlushes sync.WaitGroup
	done           chan struct{}

	// callbacks
	handleFlush FlushFunc
	handleError ErrorFunc

	// limits
	maxEntries    int
	maxRetries    int
	flushInterval time.Duration
	flushTimeout  time.Duration

	// buffer
	mu     sync.Mutex
	values []interface{}
}

// Push adds a value to the buffer.
func (b *Buffer) Push(value interface{}) {
	b.mu.Lock()
	defer b.mu.Unlock()

	b.values = append(b.values, value)

	if len(b.values) >= b.maxEntries {
		b.flush()
	}
}

// Flush flushes any pending entries asynchronously.
func (b *Buffer) Flush() {
	b.mu.Lock()
	defer b.mu.Unlock()

	b.flush()
}

// FlushSync flushes any pending entries synchronously.
func (b *Buffer) FlushSync() {
	b.mu.Lock()
	defer b.mu.Unlock()

	b.doFlush(b.values)
	b.values = nil
}

// Close flushes any pending entries, and waits for flushing to complete. This
// method should be called before exiting your program to ensure entries have
// flushed properly.
func (b *Buffer) Close() {
	b.Flush()
	close(b.done)
	b.pendingFlushes.Wait()
}

// intervalFlush starts a loop flushing at the FlushInterval.
func (b *Buffer) intervalFlush() {
	tick := time.NewTicker(b.flushInterval)
	for {
		select {
		case <-tick.C:
			b.Flush()
		case <-b.done:
			tick.Stop()
			return
		}
	}
}

// flush performs a flush asynchronuously.
func (b *Buffer) flush() {
	values := b.values
	b.values = nil

	if len(values) == 0 {
		return
	}

	b.pendingFlushes.Add(1)

	go func() {
		b.doFlush(values)
		b.pendingFlushes.Done()
	}()
}

// doFlush handles flushing of the given values.
func (b *Buffer) doFlush(values []interface{}) {
	var retries int

retry:
	ctx, cancel := context.WithTimeout(context.Background(), b.flushTimeout)

	err := b.handleFlush(ctx, values)
	cancel()

	// temporary error, retry if we haven't exceeded MaxEntries
	if e, ok := err.(temporary); ok && e.Temporary() {
		logs.Printf("temporary error flushing %d entries: %v", len(values), e)
		time.Sleep(time.Second) // TODO: backoff
		retries++
		if retries < b.maxRetries {
			goto retry
		}
		logs.Printf("max retries of %d exceeded", b.maxRetries)
	}

	if err != nil {
		b.handleError(err)
	}
}

// noopFlush function.
func noopFlush(context.Context, []interface{}) error {
	return nil
}

// noopError function.
func noopError(err error) {
	logs.Printf("unhandled error: %v", "")
}
